amaranth/hdl/_mem.py

```diff
@@ -310,10 +274,92 @@ def __init__(self, *, data_width, addr_width, domain="sync", name=None, granular
                            name=f"{name}_en", src_loc_at=1)
 
 
+class MemorySimRead:
+    def __init__(self, token, addr):
+        assert isinstance(token, MemoryToken)
+        self._token = token
+        self._addr = Value.cast(addr)
+
+    def eq(self, value):
+        return MemorySimWrite(self._token, self._addr, value)
+
+
+class MemorySimWrite:
+    def __init__(self, token, addr, data):
+        assert isinstance(token, MemoryToken)
+        self._token = token
+        self._addr = Value.cast(addr)
+        self._data = Value.cast(data)
+
+
+class MemoryToken: pass
+
+
 class MemoryInstance(Fragment):
-    def __init__(self, memory, read_ports, write_ports):
+    class _ReadPort:
+        def __init__(self, *, domain, addr, data, en, transparency):
+            assert domain is None or isinstance(domain, str)
+            if domain == "comb":
+                domain = None
+            self._domain = domain
```

Comment:
I think the amount of underscores here (within `MemoryInstance`) is *a little* excessive, but I also have no real objection to it.